File: DFS/sum_root_to_leaf_numbers.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def sumNumbers(self, root: Optional[TreeNode]) -> int:
    def dfs(node, current_number):
      if not node:
        return 0

      current_number = current_number * 10 + node.val

      if not node.left and not node.right:
        return current_number

      return dfs(node.left, current_number) + dfs(node.right, current_number)

    return dfs(root, 0)


root1 = TreeNode(1, 2, 3)
root1.left = TreeNode(2)
root1.right = TreeNode(3)
solution = Solution()
result = solution.sumNumbers(root1)
print(f'result: {result}')

File: DFS/sumRootToLeafNumbers.go
package main

import (
	"fmt"
)

// TreeNode represents a node in a binary tree.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// sumNumbers function calculates the sum of all numbers formed by root-to-leaf paths.
func sumNumbers(root *TreeNode) int {
	var dfs func(node *TreeNode, currentNumber int) int

	dfs = func(node *TreeNode, currentNumber int) int {
		if node == nil {
			return 0
		}

		currentNumber = currentNumber*10 + node.Val

		if node.Left == nil && node.Right == nil {
			return currentNumber
		}

		return dfs(node.Left, currentNumber) + dfs(node.Right, currentNumber)
	}

	return dfs(root, 0)
}

func main() {
	// Example usage:
	root := &TreeNode{Val: 1}
	root.Left = &TreeNode{Val: 2}
	root.Right = &TreeNode{Val: 3}

	result := sumNumbers(root)
	fmt.Printf("result: %d\n", result)
}
